qiskit_experiments/curve_analysis/guess.py

```diff
@@ -90,11 +108,127 @@ def frequency(
             # no oscillation signal
             return 0.0
 
-        freq_guess = max(np.abs(np.diff(y_smooth) / sampling_interval)) / (y_amp * 2 * np.pi)
+        freq_guess = max(np.abs(np.diff(y_smooth) / dt)) / (y_amp * 2 * np.pi)
 
     return freq_guess
 
 
+@_require_uniform_data
+def frequency_lorentz_fit(
+    x: np.ndarray,
+    y: np.ndarray,
+    fit_range: int = 5,
+) -> float:
+    """Get oscilaltion frequency of sinusoidal.
+
+    This function estimates frequency with FFT.
+    The FFT peak location is fine tuned by the Lorentzian fitting to give precise estimate.
+    When the estimated frequency is smaller than 150 % of a FFT frequency bin,
+    this returns the estimate by :func:`low_frequency_limit` instead.
+
+    .. note::
+
+        The offset of y data must be subtracted otherwise this may return zero frequency.
+
+    Args:
+        x: Array of x values.
+        y: Array of y values.
+        fit_range: Data points used for Lorentzian fitting.
+
+    Returns:
+        Frequency estimate.
+    """
+    dt = x[1] - x[0]
+
+    fft_y_re = np.fft.fft(np.real(y))
+    fft_y_im = np.fft.fft(np.imag(y))
+
+    fft_x = np.fft.fftshift(np.fft.fftfreq(len(x), dt))
+    fft_y = np.fft.fftshift(fft_y_re + 1j * fft_y_im)
+
+    # Use only positive side.
+    pos_inds = fft_x > 0
+    fft_x = fft_x[pos_inds]
+    fft_y = fft_y[pos_inds]
+
+    peak_ind = np.argmax(np.abs(fft_y))
+
+    # Fit a small Lorentzian to the FFT to fine tune the frequency
+    inds = list(range(max(0, peak_ind - fit_range), min(fft_x.size, peak_ind + fit_range)))
+    fit_x = fft_x[inds]
+    fit_y = fft_y[inds]
+
+    def objective(p):
+        y_complex = (p[0] + 1j * p[1]) / (1 + 1j * (fit_x - p[2]) / p[3])
+        y_out_concat = np.concatenate((y_complex.real, y_complex.imag))
+        y_ref_concat = np.concatenate((fit_y.real, fit_y.imag))
+        return y_out_concat - y_ref_concat
+
+    guess = [
+        np.real(fft_y[peak_ind]),
+        np.imag(fft_y[peak_ind]),
+        fft_x[peak_ind],
+        (fft_x[1] - fft_x[0]) / 3,
+    ]
+    fit_out, _ = optimize.leastsq(
+        func=objective,
+        x0=guess,
+    )
+    # In principle we can estimate phase from arctan of p1 and p0.
+    # However the accuracy is not quite good because of poor resolution of imaginary part
+    # due to limitation of the experimental resource, e.g. limited job circuits.
+    freq = float(fit_out[2])
+
+    if freq < 1.5 / (dt * len(x)):
```

Comment:
Maybe we can handle this outside the function, i.e. composite one, to get two individual guesses.